### ISSUE TITLE
DeepLIFT SHAP support for twin models?

### ISSUE THREAD

**Issue by jvierstra** [jmschrei/tangermeme#43]
Is there is way to get the interpretations to work for twin/siamese models. The current implementation works for a standard model, however when you make it into a twin model the convergence deltas explode. Here is a toy example:

```
class TwinModel(torch.nn.Module):
    def __init__(self, trunk):
        super().__init__()

        self.trunk = BassetTrunk()

    def forward(self, x, y):

        left = self.trunk(x)
        right = self.trunk(y)

        x = torch.subtract(left, right)

        return x

model = TwinModel(trunk)

attributions = deep_lift_shap(
    model, X0, args=(X1,), device="cpu",
    print_convergence_deltas=True,
)
```

This results in the following error:
```
tensor([0.0423, 0.1028, 0.0372, 0.1531, 0.0689, 0.2182, 0.0476, 0.1276, 0.1457,
        0.0179, 0.0019, 0.1082, 0.0898, 0.0894, 0.0438, 0.0805, 0.0792, 0.0703,
        0.0732, 0.1883], grad_fn=<AbsBackward0>)
[/home/jvierstra/.local/miniconda3/envs/tf-2.13.0/lib/python3.9/site-packages/tangermeme/deep_lift_shap.py:436](http://localhost:3001/home/jvierstra/.local/miniconda3/envs/tf-2.13.0/lib/python3.9/site-packages/tangermeme/deep_lift_shap.py#line=435): RuntimeWarning: Convergence deltas too high: tensor([0.0423, 0.1028, 0.0372, 0.1531, 0.0689, 0.2182, 0.0476, 0.1276, 0.1457,
        0.0179, 0.0019, 0.1082, 0.0898, 0.0894, 0.0438, 0.0805, 0.0792, 0.0703,
        0.0732, 0.1883], grad_fn=<AbsBackward0>)
  warnings.warn("Convergence deltas too high: " +
```

This seems to only happens when I call the trunk model twice because:

- I get an error when I call the trunk twice, but no further down stream layers
- Fails when I either give both inputs the same input tensor
- Works when I only call the trunk **once**

```
# In this case I call the trunk twice, but only output one 
# of the "arms" of the twin models

class TwinModel(torch.nn.Module):
    def __init__(self, trunk):
        super().__init__()

        self.trunk = BassetTrunk()

    def forward(self, x, y):

        left = self.trunk(x)
        right = self.trunk(y)

        #x = torch.subtract(left, right)

        return left

model = TwinModel(trunk)

attributions = deep_lift_shap(
    model, X0, args=(X1, ), device="cpu",
    print_convergence_deltas=True,
)
```
Output:
```
tensor([0.1185, 0.0172, 0.0363, 0.1596, 0.0447, 0.0583, 0.0506, 0.0168, 0.0349,
        0.0376, 0.0346, 0.0324, 0.0125, 0.0088, 0.0617, 0.0086, 0.1140, 0.0593,
        0.0843, 0.0559], grad_fn=<AbsBackward0>)
[/home/jvierstra/.local/miniconda3/envs/tf-2.13.0/lib/python3.9/site-packages/tangermeme/deep_lift_shap.py:436](http://localhost:3001/home/jvierstra/.local/miniconda3/envs/tf-2.13.0/lib/python3.9/site-packages/tangermeme/deep_lift_shap.py#line=435): RuntimeWarning: Convergence deltas too high: tensor([0.1185, 0.0172, 0.0363, 0.1596, 0.0447, 0.0583, 0.0506, 0.0168, 0.0349,
        0.0376, 0.0346, 0.0324, 0.0125, 0.0088, 0.0617, 0.0086, 0.1140, 0.0593,
        0.0843, 0.0559], grad_fn=<AbsBackward0>)
  warnings.warn("Convergence deltas too high: " +
```

**Fails** when giving both model inputs the same tensor:
```
attributions = deep_lift_shap(
    model, X0, args=(X0,), device="cpu",
    print_convergence_deltas=True,
)
```

Output:
```
tensor([0.1952, 0.2001, 0.1146, 0.0854, 0.0363, 0.1821, 0.0429, 0.1530, 0.0961,
        0.1066, 0.1031, 0.1162, 0.0694, 0.1516, 0.2144, 0.0160, 0.0655, 0.1414,
        0.1005, 0.0085], grad_fn=<AbsBackward0>)
[/home/jvierstra/.local/miniconda3/envs/tf-2.13.0/lib/python3.9/site-packages/tangermeme/deep_lift_shap.py:436](http://localhost:3001/home/jvierstra/.local/miniconda3/envs/tf-2.13.0/lib/python3.9/site-packages/tangermeme/deep_lift_shap.py#line=435): RuntimeWarning: Convergence deltas too high: tensor([0.1952, 0.2001, 0.1146, 0.0854, 0.0363, 0.1821, 0.0429, 0.1530, 0.0961,
        0.1066, 0.1031, 0.1162, 0.0694, 0.1516, 0.2144, 0.0160, 0.0655, 0.1414,
        0.1005, 0.0085], grad_fn=<AbsBackward0>)
  warnings.warn("Convergence deltas too high: " +
```

**Works** if I just call the trunk once:
```
class TwinModel(torch.nn.Module):
    def __init__(self, trunk):
        super().__init__()

        self.trunk = BassetTrunk()

    def forward(self, x, y):

        left = self.trunk(x)
        #right = self.trunk(y)

        #x = torch.subtract(left, right)

        return left

m = TwinModel(trunk)

attributions = deep_lift_shap(
    model, X0, args=(X1,), device="cpu",
    print_convergence_deltas=True,
)
```

Output deltas:
```
tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       grad_fn=<AbsBackward0>)
```

**Comment by jvierstra** [comment]
Thanks Jacob. That seemed to work. I found you also have to clone the input tensors when they contribute to both "sides" of the Siamese model. In any case I am working on small patch to the DeepLIFT code to enable multiple inputs and some dinucleotide shuffling strategies that seem to work better for variant effect interpretation.
